### PR TITLE
EventsDB: Add curator table

### LINKED ISSUE #762: Event db: Allocation curator table
This table mirrors the [StorageAllocaion.Curators](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/models.go#L601) field.

The `CuratorId` is logically a key into the user's table, however, there is no guarantee that the CuratorId will be a valid user. For example, adding an id as a curator and registering the id as a 0chain user can happen in either order.

Add the Curator table to Event database. Add code to add, and remove a curator.

We need to emit events in 0chain.storagesc to induce changes to this table.
[addCurator](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/curator.go#L99)
[removeCurator](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/curator.go#L62)

| Event db Curator | type   | StorageAllocaion.curators | type   | Note                              |
|------------------|--------|---------------------------|--------|-----------------------------------|
| allocationId     | uint   |                           |        | foreign key into allocation table |
| CuratorId        | string | curators[i]               | string |                                   |

### PR DESCRIPTION
**Pull request by amanjain97**
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master BEFORE merging this PR  
  
Associated PRs (Link as appropriate):
...

### CODE REVIEW

File: code/go/0chain.net/smartcontract/dbs/event/curator.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+type Curator struct {
+	gorm.Model
+
+	// Foreign Key
+	AllocationID string `json:"allocation_id"`
+
+	CuratorID string `json:"curator_id"`
+}
+
+func (edb *EventDb) addCurator(c Curator) error {
```
Comment (kushthedude) [inline code/go/0chain.net/smartcontract/dbs/event/curator.go]: Can we make it addOrOverwrite?

File: code/go/0chain.net/smartcontract/dbs/event/curator.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package event
```
Comment (kushthedude) [inline code/go/0chain.net/smartcontract/dbs/event/curator.go:1]: Can we add test for the same? Either unit or integration would be fine

**Comment by ssardana96** [comment]
@amanjain97 @kushthedude can we separate Block and Curator changes into 1 separate PR?
I though Block changes were done?
I see them already merged in main.

**Comment by amanjain97** [comment]
Sure let me do it by today

File: code/go/0chain.net/smartcontract/dbs/event/curator.go (hunk 1 above)
Comment (amanjain97) [reply code/go/0chain.net/smartcontract/dbs/event/curator.go]: Done

File: code/go/0chain.net/smartcontract/dbs/event/curator.go (hunk 2 above)
Comment (amanjain97) [reply code/go/0chain.net/smartcontract/dbs/event/curator.go:1]: Added.

File: code/go/0chain.net/smartcontract/dbs/event/curator.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	"fmt"
+
+	"gorm.io/gorm"
+)
+
+type Curator struct {
+	gorm.Model
+
+	// Foreign Key
+	AllocationID string `json:"allocation_id"`
```
Comment (ssardana96) [inline code/go/0chain.net/smartcontract/dbs/event/curator.go:14]: can you create a separate issue for the same

File: code/go/0chain.net/smartcontract/dbs/event/curator.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+	CuratorID string `json:"curator_id"`
```
Comment (ssardana96) [inline code/go/0chain.net/smartcontract/dbs/event/curator.go]: shouldn't this be a unique_index?

File: code/go/0chain.net/smartcontract/dbs/event/process.go [hunk 5]
```diff
@@ -35,6 +35,8 @@ const (
 	TagAddOrOverwriteWriteMarker
 	TagAddBlock
 	TagAddOrOverwriteValidator
+	TagAddCurator
```
Comment (ssardana96) [inline code/go/0chain.net/smartcontract/dbs/event/process.go]: all new tags should be added in the end
else the db will have tad ids mixed up

File: code/go/0chain.net/smartcontract/dbs/event/process.go [hunk 6]
```diff
@@ -135,6 +137,20 @@ func (edb *EventDb) addStat(event Event) error {
 			return err
 		}
 		return edb.addOrOverwriteValidator(vn)
+	case TagAddCurator:
```
Comment (ssardana96) [inline code/go/0chain.net/smartcontract/dbs/event/process.go]: TagAddOverwriteCurator

File: code/go/0chain.net/smartcontract/storagesc/curator.go [hunk 7]
```diff
@@ -115,3 +130,19 @@ func (sa StorageAllocation) isCurator(id string) bool {
...
+		CuratorID:    ci.CuratorId,
+	}
+}
+
+func emitCuratorEvent(ci *curatorInput, balances chainstate.StateContextI, txn *transaction.Transaction, eventTag event.EventTag) error {
+	data, err := json.Marshal(curatorToCuratorEvent(ci))
+	if err != nil {
+		return fmt.Errorf("failed to marshal curator: %v", err)
+	}
+	balances.EmitEvent(event.TypeStats, eventTag, txn.Hash, string(data))
```
Comment (ssardana96) [inline code/go/0chain.net/smartcontract/storagesc/curator.go]: txn_hash is not stored in data, that shouldn't be index
here ci.AllocationID should be 3rd argument index

**Review by ssardana96 — CHANGES_REQUESTED** [review]

**Comment by ssardana96** [comment]
there are couple of changes.
is this branch tested on local chain to see data is getting populated? this would be helpful to validate.

**Comment by ssardana96** [comment]
@Sriep can you please review this.
won't there be any addition of getCurator API for the same?

File: code/go/0chain.net/smartcontract/dbs/event/curator.go (hunk 3 above)
Comment (amanjain97) [reply code/go/0chain.net/smartcontract/dbs/event/curator.go:14]: Ok

**Comment by amanjain97** [comment]
```
events_db=# select * from curators;
 id |          created_at           |          updated_at           | deleted_at |                          allocation_id                           |                            curator_id                            
----+-------------------------------+-------------------------------+------------+------------------------------------------------------------------+------------------------------------------------------------------
  1 | 2022-01-26 17:24:43.13762+00  | 2022-01-26 17:24:43.142974+00 |            | fba94bd8a9fa2bd723ddbe9c08646ba3942c90ff270d356c27971a621c93c5ce | a9ae7246979694e020a24f8c88297ddfd222bc304dab4a3b90ea40c8dc2ff207
  2 | 2022-01-26 17:24:44.85506+00  | 2022-01-26 17:24:44.906015+00 |            | 441665cb639d78743ca3d7e2402638c97e2a580f00032efb555fbe18fbb9033b | 7c909d700fa97565d64af42577cec326fa4db9997a7247b41e9de3e77c3b6e4f
  3 | 2022-01-26 17:24:47.088928+00 | 2022-01-26 17:24:47.112394+00 |            | 516d6684393f5b58767123d1d0f24f241234273529e2ad9d22ec8e8524aabc8f | 699b32ffa58bb4524d276df98ac14d82b04f2d8e93b301213561cae5895890a4
  4 | 2022-01-26 17:24:49.344671+00 | 2022-01-26 17:24:49.379206+00 |            | 0f04a6b12fff1ec8da7ab3b7ead4bf99b393d3783e1a9e5a7bf5707454f7b20b | bdf248ab973d0cf2bd250f1cd42cbd80ecf07b3a037daedae857dce24c1875c1
  5 | 2022-01-26 17:24:59.702084+00 | 2022-01-26 17:24:59.732024+00 |            | 548ac15ca02413ca8b6f7dce574c6a3b463b6d77f9ddf0b3231b55dcf5a57d2c | 47d23b69b44f4fb207468dc1befa55970d26bee23673cd32165443cf1be7eb47
  6 | 2022-01-26 17:25:08.370589+00 | 2022-01-26 17:25:08.389401+00 |            | 12f9f5fd0eb3fc5846517f0e56e863f4e053f7879d1be5ef962f29593476a367 | 8ad640453597a17578eedf4e9f1fe474a9a7b7b977147d45ad227861c2ad7db9
  7 | 2022-01-26 17:28:38.662665+00 | 2022-01-26 17:28:38.70419+00  |            | 14c6a865034ff23264e7d43939680365b2670be9b2a14de973f916cd3c9df9b4 | d85c2f7521e138690afdcc16a3481d68c87f76edaacadec1bc6723fb7eb33c6a
```

I have manually added `d85c2f7521e138690afdcc16a3481d68c87f76edaacadec1bc6723fb7eb33c6a` curator id from zbox addcurator command.

File: code/go/0chain.net/smartcontract/dbs/event/curator.go (hunk 3 above)
Comment (amanjain97) [reply code/go/0chain.net/smartcontract/dbs/event/curator.go:14]: https://github.com/0chain/0chain/issues/952

**Review by kushthedude — APPROVED** [review]
Looks great! @ssardana08  We dont have any getCurator API as of now in our SCs, so I am guessing we can skip the handler . @Sriep do let know if it is needed.

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | stake_pool_update  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | sp-changes  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Pull request merged**